extensions/lifecycle/lifecycleProducer/LifecycleProducer.js

```diff
@@ -0,0 +1,256 @@
+'use strict'; // eslint-disable-line
+
+const async = require('async');
+const AWS = require('aws-sdk');
+const http = require('http');
+const Logger = require('werelogs').Logger;
+
+const BackbeatProducer = require('../../../lib/BackbeatProducer');
+const BackbeatConsumer = require('../../../lib/BackbeatConsumer');
+const AccountCredentials =
+    require('../../../lib/credentials/AccountCredentials');
+const safeJsonParse = require('../util/safeJsonParse');
+
+const CONSUMER_FETCH_MAX_BYTES = 5000020;
+
+class LifecycleProducer {
+    /**
+     * Constructor of LifecycleProducer
+     * @param {Object} zkConfig - zookeeper config
+     * @param {Object} lcConfig - lifecycle config
+     * @param {Object} s3Config - s3 config
+     * @param {String} s3Config.host - host ip
+     * @param {String} s3Config.port - port
+     * @param {Object} authConfig - auth config
+     * @param {String} authConfig.account - account name
+     * @param {Object} authConfig.vault - vault details
+     * @param {String} authConfig.vault.host - vault host ip
+     * @param {number} authConfig.vault.port - vault port
+     * @param {number} authConfig.vault.adminPort - vault admin port
+     * @param {String} transport - http or https
+     */
+    constructor(zkConfig, lcConfig, s3Config, authConfig, transport) {
+        this._log = new Logger('Backbeat:LifecycleProducer');
+        this._zkConfig = zkConfig;
+        this._lcConfig = lcConfig;
+        this._s3Client = new AWS.S3({
+            endpoint: `${transport}://${s3Config.host}:${s3Config.port}`,
+            credentials: new AccountCredentials(authConfig, this._log),
```

Comment:
When using vault, we will have to use the admin credentials in admin-backbeat.json to create new access keys for the account owning the bucket, and cache them in memory, like echo feature is doing (it also creates accounts on the destination but we do not want this for lifecycle). See how the `VaultClientCache` class is used to do this.

A corollary is that we'll need different S3 clients for each account, each time providing them the proper account access/secret key for the object/bucket to process. For replication, we create new clients for every object to replicate, sounds a bit overkill but does not seem to cause major CPU/memory issues in practice so I think it's acceptable.

Reply:
I've updated this to generate a new key for each unique canonical ID (using cache if the canonical ID has been seen before). I then create a new S3 client for each `getBucketLifecycle` request using the access/secret keys.